Final: first version.

JadrankoLucic opened the pull request:
Kritike i komentari dobrodošli.

Commit:
Final: first version.

ncaklovic commented:
Svi testovi prolaze :-), to je dobro. 
Možeš ih pisati kraće, poput
```
    block_processor bp;
    bp.add_block(block_factory::Create("add", { 5 }));
    Assert::AreEqual(10., bp.calc_all(5.));
```

Nema calc_all sa više vrijednosti (iz ulaznog filea, zadnji dio zadatka)?

JadrankoLucic commented:
1. dodat ću još testova
2. smetnuo sam (ustvari nisam niti gledao) task: "unijeti naziv tekstualne datoteke sa nizom vrijednosti - izvršit će se svi blokovi nad svim vrijednostima i snimiti izlazni file sa rezultatima". To ću isto napraviti.
3. dodat ću još blokova (možda)

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+{
+  std::string arguments;
+
+  std::for_each(parameters_.begin(), parameters_.end(), [&](double const &par) {
```
std::accumulate?

JadrankoLucic replied:
promijenio

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+
+}
+
+void block_processor::add_block(std::unique_ptr<block> &b)
```
preuzima lifetime objekta, jasnije je ako se posalje by-value, u tom slucaju caller mora napraviti move

void block_processor::add_block(std::unique_ptr<block> b)

JadrankoLucic replied:
Primijenio pravilo: Take a unique_ptr<obj> parameter to express that a function assumes ownership of a obj
http://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+
+double block_processor::calc_all(double input) const
+{
+  double calculated_value = input;
```
std::accumulate?

JadrankoLucic replied:
changed

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+  double calculated_value = input;
+  for (auto const& b : blocks_)
+  {
+    calculated_value = (*b).calc(calculated_value);
```
(*b).calc(calculated_value); ekvivalentno kao b->calc(calculated_value)

JadrankoLucic replied:
meni je ovako jasnije

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+  temp_vector.reserve(blocks_.size());
+  std::transform(blocks_.begin(), blocks_.end(), std::back_inserter(temp_vector), [](std::unique_ptr<block> const &b)
+  {
+    return (*b).get_command();
```
b->get_command()

JadrankoLucic replied:
meni je ovako jasnije

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+  return temp_vector;
+}
+
+void block_processor::load(sequence_list sl)
```
sequence_list je definiran kao std::vector<std::pair<std::string, std::vector<double>>>
pozivom se dogada kopija

void block_processor::load(const sequence_list& sl)

JadrankoLucic replied:
changed.

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+{
+  blocks_.clear();
+  for (sequence_list::iterator it = sl.begin(); it != sl.end(); ++it) {
+    const std::string block_name = (*it).first;
```
it->first

JadrankoLucic replied:
meni je ovako jasnije.

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+    const std::string block_name = (*it).first;
+    const std::vector<double> parameter_values = it->second;
+    std::unique_ptr<block> bl = block_factory::Create(block_name, parameter_values);
+    blocks_.push_back(std::move(bl));
```
block_processor::add_block ?

JadrankoLucic replied:
changed.

jan-kelemen commented on final/tool/block.cpp:
```diff
@@ -0,0 +1,128 @@
+  std::vector<std::string> temp_vector;
+  temp_vector.reserve(create_methods_.size());
+
+  //First solution: Copy all value fields from map to a vector using transform() & Lambda function
```
ovo je bilo jasnije, std::for_each s [&] u capture listi lambde moze imati proizvoljnih side efekata

JadrankoLucic replied:
Uzeo 'First solution'

jan-kelemen commented on final/tool/block.h:
```diff
@@ -0,0 +1,72 @@
+{
+public:
+  using TBlockCreateMethod = std::unique_ptr<block>(*)(const std::vector<double>& parameter);
+  block_meta(std::string description = "", TBlockCreateMethod block_create_method = nullptr)
```
**explicit** block_meta(std::string description = "", TBlockCreateMethod block_create_method = nullptr);

https://en.cppreference.com/w/cpp/language/explicit

konstruktori definirani s jednim parametrom, sto ovaj je ukoliko se za drugi argument uzme defaultni omogucavaju implicitne konvezije, sto pretpostavljam da ovdje nije zelja

npr, recimo da postoji funkcija:
void foo(block_meta meta);

bez explicit, ovo je moguce pozvati s:
std::string pero = "pero"
foo(pero);

sto pretpostavljam da nije zeljeno

JadrankoLucic replied:
done.

jan-kelemen commented on final/tool/block.h:
```diff
@@ -0,0 +1,72 @@
+  using TBlockCreateMethod = std::unique_ptr<block>(*)(const std::vector<double>& parameter);
+  block_meta(std::string description = "", TBlockCreateMethod block_create_method = nullptr)
+  {
+    description_ = description;
```
inicijalizacija u initializer-listi, 

block_meta(std::string description = "", TBlockCreateMethod block_create_method = nullptr) : description_{std::move(description)}, block_create_method_{block_create_method} 
{ ; }

JadrankoLucic replied:
promijenio.

jan-kelemen commented on final/tool/block_lim.cpp:
```diff
@@ -0,0 +1,24 @@
+double lim::calc(double input)
+{
+  is_blocktype_registered_;
+  double ret = input;
```
std::clamp?

JadrankoLucic replied:
primijenio

jan-kelemen commented on final/tool/block_add.cpp:
```diff
@@ -0,0 +1,19 @@
+  return initial_value_ + input;
+};
+
+std::string self_registered_in_factory<add>::description_ = "Summarize input with parameter value. Example: add 5";
```
mozda dodati konstruktor u 

self_registered_in_factory(const char* description, const char* name) pa ne treba svaki blok definirati svoje metode?
disklejmer: ovaj prijedlog podrazumijeva da su uvijek inicijalizirani u compile-timeu https://stackoverflow.com/questions/2001286/const-char-s-in-c

------
inicijalizaciju statica je moguce izbjeci tako da svaka klasa dogovorno ima const char* member description_ pa iskoristiti CRTP https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern
static std::string self_registered_in_factory::GetDescription() { return TBlock::description_; }

JadrankoLucic replied:
first part not accepted.
the second part is partially accepted.

jan-kelemen commented on final/tool/menus.cpp:
```diff
@@ -0,0 +1,192 @@
+    std::cout << caption;
+    getline(std::cin, input);
+    // This code converts from string to number safely.
+    std::stringstream input_ss(input);
```
https://en.cppreference.com/w/cpp/string/basic_string/stol
https://en.cppreference.com/w/cpp/string/basic_string/stof

JadrankoLucic replied:
with current code I have 'try' approach instead of handling exception

jan-kelemen commented on final/tool/menus.cpp:
```diff
@@ -0,0 +1,192 @@
+  system("cls");
+  std::string block_list_menu = "Available block types:\n";
+  std::vector<std::string> block_list = block_factory::GetBlockTypes();
+  for (std::vector<std::string>::iterator it = block_list.begin(); it != block_list.end(); ++it)
```
std::accumulate?

JadrankoLucic replied:
done

jan-kelemen commented on final/tool/menus.cpp:
```diff
@@ -0,0 +1,192 @@
+  //show block sequence
+  system("cls");
+  std::cout << "Block sequence:\n" << std::endl;
+  int index = 0;
```
```
std::vector<std::string> temp_v = bp.get_sequence();
std::for_each(temp_v.begin(), temp_v.end(), [index = 0](std::string const &c) mutable { index++;  std::cout << index << ":\t" << c << std::endl; });
```

u principu, sto vise smanjiti scope lambde, lambda sa capture listom [&] lovi svakakve okolne varijable poglavlje "Lambda capture" https://en.cppreference.com/w/cpp/language/lambda

JadrankoLucic replied:
done.

jan-kelemen commented on final/tool/menus.cpp:
```diff
@@ -0,0 +1,192 @@
+  std::cout << "Block sequence:\n" << std::endl;
+  int index = 0;
+  std::vector<std::string> temp_v = bp.get_sequence();
+  std::for_each(temp_v.begin(), temp_v.end(), [&](std::string const &c)
```
ili samo obican for kada je vec potreban index?

JadrankoLucic replied:
see above

jan-kelemen commented on final/tool/menus.cpp:
```diff
@@ -0,0 +1,192 @@
+      block_name_taken = true;
+    }
+    else
+      parameters.push_back(atof(value.c_str()));
```
std::stof(value)

JadrankoLucic replied:
done.

jan-kelemen reviewed (commented):
review na zahtjev od @JadrankoLucic

Commit:
Final:
- process inputs from file
- more tests
- some refactoring
- new methods on block_processor: load_from_stream, bulk_calc, get_block_command_from_text

Commit:
Jan's comments are resolved.
New tests
Refactoring
